java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -242,21 +243,44 @@ public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
 
     @Override
     public void visitBinaryExpression(BinaryExpressionTree tree) {
-      if(tree.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
-        increaseComplexityByOne(tree.operatorToken());
-        ExpressionTree left = ExpressionUtils.skipParentheses(tree.leftOperand());
-        if(left.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) left).operatorToken());
-          complexity--;
-        }
-        ExpressionTree right = ExpressionUtils.skipParentheses(tree.rightOperand());
-        if(right.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) right).operatorToken());
-          complexity--;
+      if (tree.is(CONDITIONAL_AND, CONDITIONAL_OR) && !ignored.contains(tree)) {
+        List<SyntaxToken> flatOperators = new ArrayList<>();
+        flattenLogicalExpression(0, flatOperators, tree);
+
+        // first binary expression
+        increaseComplexityByOne(flatOperators.get(0));
+
+        // rest
+        for (int i = 1; i < flatOperators.size(); i++) {
+          SyntaxToken previous = flatOperators.get(i - 1);
+          SyntaxToken current = flatOperators.get(i);
+          if (!sameOperators(previous, current)) {
+            increaseComplexityByOne(current);
+          }
         }
       }
       super.visitBinaryExpression(tree);
     }
+
+    private static boolean sameOperators(SyntaxToken op1, SyntaxToken op2) {
+      return op1.text().equals(op2.text());
+    }
+
+    private void flattenLogicalExpression(int i, List<SyntaxToken> operators, ExpressionTree expression) {
```

Comment:
IMO we could try to write a small method that maps a binary exp to a stream of tree then flatmap this recursively to obtain what's desired in a more functional style avoiding to deal with index.